File: integrations/alga-agent-sdk-py/alga_agent_sdk/dedup.py
from __future__ import annotations

import asyncio
import time


class MessageDedup:
    def __init__(self, max_entries: int = 1000, ttl_seconds: float = 300.0):
        self._max = max(1, max_entries)
        self._ttl = ttl_seconds if ttl_seconds > 0 else 300.0
        self._seen: dict[str, float] = {}
        self._lock = asyncio.Lock()

    async def is_duplicate(self, message_id: str) -> bool:
        now = time.monotonic()
        async with self._lock:
            self._evict(now)
            if len(self._seen) >= self._max:
                self._trim_to(self._max * 9 // 10)
            if message_id in self._seen:
                self._seen[message_id] = now
                return True
            self._seen[message_id] = now
            return False

    async def clear(self) -> None:
        async with self._lock:
            self._seen.clear()

    @property
    def size(self) -> int:
        return len(self._seen)

    def _evict(self, now: float) -> None:
        for k, t in list(self._seen.items()):
            if now - t > self._ttl:
                del self._seen[k]

    def _trim_to(self, target: int) -> None:
        target = max(target, 0)
        while len(self._seen) > target:
            oldest_key = None
            oldest_time = 0.0
            first = True
            for k, t in self._seen.items():
                if first or t < oldest_time:
                    oldest_time = t
                    oldest_key = k
                    first = False
            if oldest_key is None:
                break
            del self._seen[oldest_key]

File: integrations/alga-agent-sdk-rs/src/dedup.rs
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};

pub struct MessageDedup {
    seen: Mutex<HashMap<String, Instant>>,
    max: usize,
    ttl: Duration,
}

impl MessageDedup {
    pub fn new(max_entries: usize, ttl: Duration) -> Self {
        let max = max_entries.max(1);
        let ttl = if ttl.is_zero() { Duration::from_secs(300) } else { ttl };
        Self {
            seen: Mutex::new(HashMap::new()),
            max,
            ttl,
        }
    }

    pub fn is_duplicate(&self, message_id: &str) -> bool {
        let mut seen = self.seen.lock().expect("dedup mutex poisoned");
        let now = Instant::now();

        seen.retain(|_, ts| now.duration_since(*ts) < self.ttl);

        if seen.len() >= self.max {
            let target = (self.max * 9) / 10;
            Self::trim_to(&mut seen, target);
        }

        if seen.contains_key(message_id) {
            seen.insert(message_id.to_string(), now);
            return true;
        }

        seen.insert(message_id.to_string(), now);
        false
    }

    pub fn clear(&self) {
        self.seen.lock().expect("dedup mutex poisoned").clear();
    }

    pub fn size(&self) -> usize {
        self.seen.lock().expect("dedup mutex poisoned").len()
    }

    fn trim_to(seen: &mut HashMap<String, Instant>, target: usize) {
        let target = target.min(seen.len());
        while seen.len() > target {
            let oldest = seen
                .iter()
                .min_by_key(|(_, ts)| *ts)
                .map(|(k, _)| k.clone());
            match oldest {
                Some(key) => {
                    seen.remove(&key);
                }
                None => break,
            }
        }
    }
}
